### PR TITLE
Add external access support in streamlit

### LINKED ISSUE #1433: SNOW-1621993: Ability to provide external access integration and secrets when deploying the streamlit
### Description

Streamlit applications often require integration with external services like OpenAI and Jira. Currently, Snow-CLI lacks the ability to provide this, limiting its functionality for building comprehensive data applications.

### Context

My streamlit requires to an external access integration to do some API calls in Jira. This feature will be very helpful as I do not have to alter the streamlit manually to grant external access to the streamlit.

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [x] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
Added support for external access in streamlit. Closes #1433

### CODE REVIEW

File: src/snowflake/cli/api/sql_execution.py [hunk 1]
```diff
@@ -35,16 +35,20 @@
...
-    def __init__(self):
+    def __init__(self, connection: SnowflakeConnection | None = None):
         self._snowpark_session = None
+        self._connection = connection
 
     @property
-    def _conn(self):
+    def _conn(self) -> SnowflakeConnection:
+        if self._connection:
+            return self._connection
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/sql_execution.py:51]: This change makes the testing much more easier.

File: src/snowflake/cli/api/project/schemas/entities/common.py [hunk 2]
```diff
@@ -109,3 +109,29 @@ def get_type(self) -> type:
...
+        default=[],
+    )
+    secrets: Optional[Dict[str, str]] = Field(
+        title="Assigns the names of secrets to variables so that you can use the variables to reference the secrets",
+        default={},
+    )
+
+    def get_external_access_integrations_sql(self):
+        external_access_integration_name = ", ".join(
+            f"{e}" for e in self.external_access_integrations
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/api/project/schemas/entities/common.py:131]: Is this needed? You do List[str] from List[str]? 🤔

File: tests_integration/test_data/projects/streamlit_v2_external_access/snowflake.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    identifier:
+      name: test_streamlit_deploy_snowcli_ext_access
+    title: "My Fancy Streamlit"
+    stage: streamlit
+    query_warehouse: xsmall
+    main_file: streamlit_app.py
+    external_access_integrations:
+      - snowflake_docs_access_integration
+    secrets:
+      generic_secret: external_access_db.public.test_secret
```
Comment (sfc-gh-pczajka) [inline tests_integration/test_data/projects/streamlit_v2_external_access/snowflake.yml:14]:
```suggestion
      generic_secret: snowcli_db.public.test_secret
```

**Review by sfc-gh-astus — APPROVED** [review]

**Review by sfc-gh-pczajka — APPROVED** [review]

**Review by sfc-gh-pczajka — APPROVED** [review]

**Review by sfc-gh-astus — APPROVED** [review]

File: tests_integration/test_data/projects/streamlit_v2_external_access/snowflake.yml (hunk 3 above)
Comment (sfc-gh-astus) [reply tests_integration/test_data/projects/streamlit_v2_external_access/snowflake.yml:14]: I will replace `external_access_db` with `snowcli_db` in my next PR.

**Pull request merged**